### PR TITLE
Enable CI

### PR DESCRIPTION
**Pull request by ravimeda**
Fixes https://github.com/dotnet/dotnet-buildtools-prereqs-docker/issues/2

- `Dockerfile.linux.imagebuilder` provides an image that has ImageBuilder and the repo
- `build.ps1` runs in a container with `Dockerfile.linux.imagebuilder` image
- CI runs `build.ps1` for each distro

### CODE REVIEW

**Comment by ravimeda** [comment]
@dotnet-bot test CI please

File: Dockerfile.linux.imagebuilder [hunk 1]
```diff
@@ -0,0 +1,4 @@
+FROM microsoft/dotnet-buildtools-prereqs:image-builder-jessie-20180117125404
```
Comment (MichaelSimons) [inline Dockerfile.linux.imagebuilder]: I think a usage note should be added similar to the one Dockerfile.linux.testrunner has.

File: Dockerfile.linux.imagebuilder [hunk 2]
```diff
@@ -0,0 +1,4 @@
...
+
+WORKDIR /
```
Comment (MichaelSimons) [inline Dockerfile.linux.imagebuilder]: I don't think we should be coping the contents into the root folder.  I think a subfolder would be more appropriate (should align with Dockerfile.linux.testrunner)

File: Dockerfile.linux.testrunner [hunk 3]
```diff
@@ -0,0 +1,7 @@
+# Use this Dockerfile to create a Linux testrunner image
+# Usage: docker run --rm -v /var/run/docker.sock:/var/run/docker.sock testrunner pwsh -File build.ps1 <params....>
+
+FROM microsoft/dotnet-buildtools-prereqs:debian-stretch-docker-testrunner-63f2145-20184325094343
+
+WORKDIR src
```
Comment (MichaelSimons) [inline Dockerfile.linux.testrunner]: This directory should be rooted

File: build.ps1 [hunk 4]
```diff
@@ -34,15 +33,32 @@ function ExecuteWithRetry($Command) {
...
-    -v "${PSScriptRoot}:/repo" `
-    -w /repo `
-    $ImageBuilderImageName `
-    build --manifest "manifest.json" --path "$DockerfilePath" "$ImageBuilderCustomArgs"
+Invoke-CleanupDocker
 
-if ($LastExitCode -ne 0) {
-    throw "Failed executing ImageBuilder."
+try {
+    ExecuteWithRetry docker pull 'microsoft/dotnet-buildtools-prereqs:image-builder-jessie-20180117125404'
```
Comment (MichaelSimons) [inline build.ps1]: We can't have the image-builder name duplicated in two spots.  This is going to get out of sync with the Dockerfile.

File: netci.groovy [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+def machineLabel = 'latest-or-auto-docker'
+def distroList = ['alpine', 'centos', 'debian', 'fedora', 'opensuse', 'ubuntu']
+
+distroList.each { distro ->
+    def newJobName = Utilities.getFullJobName(project, "${distro}", isPR)
+
+    def newJob = job(newJobName) {
+        steps {
+            shell("docker build --rm -t testrunner -f ./Dockerfile.linux.testrunner . ")
+            shell("docker run -v /var/run/docker.sock:/var/run/docker.sock testrunner pwsh -NoProfile -File ./build.ps1 -DockerfilePath \"*${distro}*\" -CleanupDocker")
```
Comment (MichaelSimons) [inline netci.groovy]: I think it would be better to use a more precise path `/src/${distro}/*`

File: netci.groovy [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+        }
+    }
+
+    Utilities.setMachineAffinity(newJob, hostOS, machineLabel)
+    Utilities.standardJobSetup(newJob, project, isPR, "*/${branch}")
+    Utilities.addGithubPRTriggerForBranch(newJob, branch, "${distro} Dockerfiles")
```
Comment (MichaelSimons) [inline netci.groovy]: `${distro}` would be shorter to type when wanting to trigger a job and would also better align with the job name.

File: Dockerfile.linux.imagebuilder (hunk 1 above)
Comment (ravimeda) [reply Dockerfile.linux.imagebuilder]: Ok. How about -

```
# Use this Dockerfile to create an image containing this repo and ImageBuilder
# Usage: docker build -t imagebuilder -f ./Dockerfile.linux.imagebuilder .
```

File: netci.groovy (hunk 5 above)
Comment (ravimeda) [reply netci.groovy]: Good point.  Updated as `-DockerfilePath \"src/${distro}/*\"`

File: netci.groovy (hunk 6 above)
Comment (ravimeda) [reply netci.groovy]: Ok.  Updated as `Utilities.addGithubPRTriggerForBranch(newJob, branch, "${distro}")`

**Comment by ravimeda** [comment]
As discussed offline, let's separate the changes to `build.ps1` so that changes in this PR are CI specific.
Hence submitted https://github.com/dotnet/dotnet-buildtools-prereqs-docker/pull/5

**Comment by ravimeda** [comment]
@dotnet-bot test CI please

**Comment by ravimeda** [comment]
@dotnet-bot test ci please

File: netci.groovy [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+def hostOS = "Ubuntu16.04"
+def machineLabel = 'latest-or-auto-docker'
+def distroList = ['alpine', 'centos', 'debian', 'fedora', 'opensuse', 'ubuntu']
+
+distroList.each { distro ->
+    def newJobName = Utilities.getFullJobName(project, "${distro}", isPR)
+
+    def newJob = job(newJobName) {
+        steps {
+            shell("docker build -t runner --pull -f ./Dockerfile.linux.runner . ")
```
Comment (MichaelSimons) [inline netci.groovy]: Nit - there is an unnecessary space after the ending `.`

Comment (ravimeda) [reply netci.groovy]: Fixed.

**Comment by ravimeda** [comment]
@dotnet-bot test ci please

**Comment by mmitche** [comment]
@dotnet-bot test ci please

**Comment by mmitche** [comment]
@dotnet-bot test ci please

**Comment by ravimeda** [comment]
@dotnet-bot test ci please

**Comment by ravimeda** [comment]
@dotnet-bot test ci please

**Comment by ravimeda** [comment]
@dotnet-bot test ci please

**Comment by ravimeda** [comment]
@dotnet-bot test ci please

**Comment by ravimeda** [comment]
@dotnet-bot test ci please

File: netci.groovy [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+def project = GithubProject
+def branch = GithubBranchName
+def isPR = true
+def distroList = ['alpine', 'centos/6', 'centos/7', 'debian', 'fedora', 'opensuse', 'ubuntu/14', 'ubuntu/16', 'ubuntu/17']
+
+distroList.each { distro ->
+    def(distroName, distroVersion) = distro.tokenize('/')
+    if (distroVersion != null) {
+        distroName = distroName.concat("_" + distroVersion)
+    }
```
Comment (MichaelSimons) [inline netci.groovy]: The logic to replace `/` with `_` seems overly complicated.  Why not simply call replace?  `distro = distro.replace('/', '_')

Comment (ravimeda) [reply netci.groovy]: Yes, `distro = distro.replace('/', '_')` works too.

Comment (MichaelSimons) [reply netci.groovy]: It is more than about functionality though - it is readability and clarity.  The current pattern complicates things and almost makes the reader think there is more to what is really being done.

Comment (ravimeda) [reply netci.groovy]: Resolved through https://github.com/dotnet/dotnet-buildtools-prereqs-docker/pull/14

**Pull request merged**